keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,430 @@
+```
+
+No change from current Kubernetes behavior:
+
+1. Kubelet does not see any `SELinuxRelabelPolicy` configured in the pod and thus mounts `myclaim` PVC as usual and if the underlying volume supports SELinux, it passes it to the container runtime with ":Z".
+   Kubelet passes also implicit Secret volume with token with ":Z".
+2. Container runtime allocates a new unique SELinux label to the pod and recursively relabels all volumes with ":Z" to this label. 
+
+
+
+#### Story 2
+
+User (or something else, e.g. an admission webhook) configures SELinux label for a pod.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testpod
+spec:
+  securityContext:
+    seLinuxOptions:
+      level: s0:c10,c0
+  containers:
+    - image: nginx
+      name: nginx
+      volumeMounts:
+        - name: vol
+          mountPath: /mnt/test
+  volumes:
+      - name: vol
+        persistentVolumeClaim:
+          claimName: myclaim
+```
+
+No change from current Kubernetes behavior:
+
+1. Kubelet does not see any `SELinuxRelabelPolicy` configured in the pod and thus mounts `myclaim` PVC as usual and if the underlying volume supports SELinux, it passes it to the container runtime with ":Z".
+   Kubelet passes also implicit Secret volume with token with ":Z".
+2. Container runtime uses SELinux label "s0:c10,c0", as instructed by Kubernetes. It will recursively relabels all volumes with ":Z" to this label.
+
+#### Story 3
+
+User (or something else, e.g. an admission webhook) configures SELinux label for a pod.
+User chooses `SELinuxRelabelPolicy: "OnVolumeMount"`, because they expect a potentially large volume to be used by the pod.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testpod
+spec:
+  securityContext:
+    seLinuxOptions:
+      level: s0:c10,c0
+    seLinuxRelabelPolicy: OnVolumeMount
+  containers:
+    - image: nginx
+      name: nginx
+      volumeMounts:
+        - name: vol
+          mountPath: /mnt/test
+  volumes:
+      - name: vol
+        persistentVolumeClaim:
+          claimName: myclaim
+```
+
+In this case, kubelet tries to mount all pod's volumes with `-o context=s0:c10,c0` mount option`.
+If it succeeds, it passes the volume to the container runtime without ":Z" and the container runtime does not relabel the volume.
+See [New Kubernetes behavior](#new-kubernetes-behavior) for error cases.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+* Unit tests:
+   * API validation (all permutations missing / present PodSecurityPolicy.SELinuxOptions & SELinuxRelabelPolicy & container.SecurityPolicy.SELinuxOptions)
+   * Passing mount options from kubelet to volume plugins.
+* E2e tests:
+   * Check no recursive `chcon` is done on a volume when not needed /
+   * Check recursive `chcon` is done on a volume when needed (with a matrix of SELinuxOptions / SELinuxRelabelPolicy).
+* Prepare e2e job that runs with SELinux in Enforcing mode!
```

Comment:
We still lack ability to test selinux features in e2e IIRC

Reply:
I think we need an e2e job to test this feature, so someone has to create such a job (and it may not be easy).